rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,153 @@
+# RFC 9292 - 2021-09-21 - `throttle` transform
+
+This RFC proposes the addition of a new transform that provides a user the ability to control the throughput of specific event streams.
+
+## Context
+
+- [#258](https://github.com/vectordotdev/vector/issues/258)
+
+## Scope
+
+### In scope
+
+- Dropping events (logs and metrics) to rate limit an event stream
+- Rate limit by both bytes (unserialized) and events
+- Exclude events with VRL conditions
+- Specify buckets based on the value of a key in the event
+
+### Out of scope
+
+- Apply backpressure as a form of rate limiting
+
+## Pain
+
+- Users lack the necessary tooling to control throughput which can cause excessive costs and negatively impact downstream services due to increased load
+- Admins cannot set quotas for users/usergroups utilizing Vector
+
+## Proposal
+
+### User Experience
+
+The `throttle` transform can be used to rate limit specific subsets of your event stream to limit load on downstream services or to enforce quotas on users.
+You can enforce rate limits on events or their raw size, as well as excluding events based on a VRL condition to avoid dropping critical logs. Rate limits
+can be applied globally across all logs or by specifying a key to create buckets of events to rate limit more granularly.
+
+The initial implementation will shed load by dropping any events beyond the configured rate limit. This could be configured in the future to apply backpressure 
+or connected to Vector's dead-letter pipelines feature.
+
+### Implementation
+
+The `throttle` transform will leverage the existing [Governor](https://docs.rs/governor/0.3.2/governor/index.html) crate. We will expose the bytes or events
+per second as user-facing configuration (or expand to more time frames) and translate that into a `governor::Quota` for internal use. We should not expose
+the `Quota` directly to users to ensure ease of configuration for end users.
+
+Config:
+
+```rust
+pub struct ThrottleConfig {
+    // One of events_* or bytes_* may be configured, but not both
+    events_per_second: u32,
+    bytes_per_second: u32,
```

Comment:
Just pulling over my comment from the implementation RFC. I think we should allow configurable windows. There is crossover with this feature and what we've been calling "quotas". In the case of quota, users will want to throttle longer windows, such as hours or days. This allows for more fluctuation but ensures that bandwidth is not exceeded.

Overall, we might want to step back and discuss how this feature fits into the larger story of rate limiting and quotas, something a brief should do.

Reply:
Yeah, agreed, I think just per-second is a bit limiting.

Reply:
Can definitely play around with either an enum of time frames (s/m/h) or arbitrary, at least at the time I was writing the spike per second was the most easily understood for me testing usage.

I definitely see the value and agree that larger time periods would be valuable and/or required.

Reply:
@binarylogic should I update the spike/rfc that the configuration parameters mirror the "characteristics" from the `Vector Durability` document?